File: workspace/busking-gps-mapper/web/adapter/inbound/join_driving_controller.go
package web_inbound

import (
	"net/http"

	"busking.org/gps-mapper/domain/application/model"
	"busking.org/gps-mapper/domain/application/service"
	"github.com/gin-gonic/gin"
)

type JoinDrivingRequestBody struct {
	Bus         *BusIdObject `json:"bus" binding:"required"`
	Destination int          `json:"destination" binding:"required"`
}

func JoinDrivingController(webAdapter *gin.Engine, drivingService *service.DrivingService) {
	const PATH = "/api/realtime/driving/join"
	webAdapter.POST(PATH, func(ctx *gin.Context) {
		var payload JoinDrivingRequestBody
		if err := ctx.ShouldBindJSON(&payload); err != nil {
			ctx.AbortWithStatusJSON(http.StatusBadRequest, map[string]any{
				"status":  "400 bad request",
				"message": err.Error(),
			})
			return
		}

		cmd := &service.JoinDrivingCommand{
			BusId: model.BusId{
				CompanyId: *payload.Bus.CompanyId,
				No:        *payload.Bus.No,
			},
			Destination: payload.Destination,
		}
		if err := drivingService.JoinDriving(cmd); err != nil {
			ctx.AbortWithStatusJSON(http.StatusBadRequest, map[string]any{
				"status":  "400 bad request",
				"message": err.Error(),
			})
			return
		}

		ctx.AbortWithStatusJSON(http.StatusOK, map[string]any{
			"status": "200 ok",
		})
	})
}

File: workspace/busking-front-server/src/main/java/org/comfort42/busking/web/adapter/inbound/JoinDrivingController.java
package org.comfort42.busking.web.adapter.inbound;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.comfort42.busking.web.adapter.inbound.schema.BusIdObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;

import java.util.Objects;

@RestController
@RequestMapping("/api/realtime/driving")
@RequiredArgsConstructor
class JoinDrivingController {

    record JoinDrivingRequestBody(BusIdObject bus, long destination) {
    }

    private final ObjectMapper objectMapper;

    @PostMapping("/join")
    ResponseEntity<?> joinDriving(
            @Value("${busking.gps-mapper.url}") final String gpsMapperUrl,
            @RequestBody final JoinDrivingRequestBody payload) throws JsonProcessingException {

        if (payload.bus() == null) {
            return ResponseEntity
                    .badRequest()
                    .contentType(MediaType.APPLICATION_JSON)
                    .body("{\"status\": \"400 bad request\", \"message\": \"$.busId is required\"}");
        }

        if (payload.destination() < 0) {
            return ResponseEntity
                    .badRequest()
                    .contentType(MediaType.APPLICATION_JSON)
                    .body("{\"status\": \"400 bad request\", \"message\": \"$.destination must be positive\"");
        }

        try {
            return WebClient.create()
                    .post()
                    .uri(gpsMapperUrl + "/api/realtime/driving/join")
                    .contentType(MediaType.APPLICATION_JSON)
                    .bodyValue(objectMapper.writeValueAsString(payload))
                    .retrieve()
                    .toEntity(String.class)
                    .block();
        } catch (WebClientResponseException e) {
            final var statusCode = e.getStatusCode();
            return ResponseEntity
                    .status(statusCode)
                    .contentType(Objects.requireNonNull(e.getHeaders().getContentType()))
                    .body(e.getResponseBodyAsString());
        }

    }

}
